Features/cgu/25.2/table organizer form 389574

cguglielmo opened the pull request.

cguglielmo committed:
Table: add table organizer form

The form is a JavaScript implementation of OrganizeColumnsForm.java.
Unlike the Java form, it is only added to the table of PageWithTable,
other tables don't have the menu by default.

389574

cguglielmo committed:
Table: add column width group to header menu

Also:
- The text of the header menu group is now reset to the text of the
  focused button and not always to the default text.
- Consolidated duplicated code for setting ARIA rules.

204393

cguglielmo committed:
NumberField: accept input on ENTER

Normally, acceptInput() is called on field blur or if a keystroke is
executed. Because the number field does not have any keystroke,
pressing ENTER does not accept the input. If the field is used
on a form with an Ok/Save button that is triggered by ENTER, it
normally doesn't matter. But if there is no such button, e.g. on a
popup, ENTER should accept the input so the user doesn't need to press
TAB.

This already works correctly for other value fields.

333898

cguglielmo committed:
Simplify Organize Columns Form

- Remove columns indicating sort, filter, grouping and customizing state
  They make it look messy. The columns already show their
  states, that should be sufficient.
- Remove menus for sorting, filtering, grouping have been removed
  They make it look messy. They are not really needed as the actions can
  be executed on the columns directly.
- Remove width column
  The columns themselves will show the width in their header menu.
- Remove CopyWidthsOfColumnsMenu
  Widths can be copied from the new column directly.
- Reduce form width
  There are fewer menus and columns on the right, it can be smaller.
- Move move-menus to the right
  They are on the right on other tables as well.
- Don't hide move-menus if they cannot be used on the selected rows.
  For example: if last row is selected, move down menu was invisible.
  After moving row up, it gets visible and the move up menu suddenly is
  at another position.
- Add tooltips for icon-only-menus
- Select profile cell text when profile is added / edited
- Deselect profile table rows if a colum row is selected and vice
  versa.

389580

cguglielmo committed:
Table: add column should be visible even if there are fixed columns

fschinkel commented on eclipse-scout-core/src/menu/menus.ts:
```diff
@@ -203,5 +203,32 @@ export const menus = {
+    if (menusToDelete.length === 0) {
+      return;
+    }
+    let menus = menuOwner.menus.slice();
```
you can use `arrays.diff`

fschinkel commented on eclipse-scout-core/src/menu/menus.ts:
```diff
@@ -203,5 +203,32 @@ export const menus = {
+    }
+    let menus = menuOwner.menus as ObjectOrChildModel<Menu>[];
+    menuOwner.setMenus(menus.concat(menusToInsert));
+  },
```
do we need to check if menus are already part of the `MenuOwner`

cguglielmo replied:
The menu owner already does it, if it is required. I don't think we should do it here

fschinkel commented on eclipse-scout-core/src/table/organizer/TableOrganizer.ts:
```diff
@@ -121,28 +114,50 @@ export class TableOrganizer implements ObjectWithType {
+    // Add guiOnly columns to visibleColumns array if they are not already included
+    // moveColumn only works with indices based on all visibleColumns including guiOnly columns
+    let guiOnlyColumns = this.table.columns.filter(column => column.guiOnly);
+    for (const column of guiOnlyColumns.reverse()) {
```
`Table.filterColumns`?

fschinkel commented on eclipse-scout-core/src/table/organizer/TableOrganizerForm.ts:
```diff
@@ -0,0 +1,199 @@
+  table: Table;
+  profilesTable: ProfilesTable;
+  columnsTable: ColumnsTable0;
+  keyColumn: Column<Column>;
```
I think we do not need to export the whole ShowInvisibleColumnsFormModel that leads to this strange name. Can you fix clean up that file (move interfaces to main widget and remove json model from index.ts)?

cguglielmo replied:
You are right, this is the current pattern. But unfortunately if we did this, it would make extending from that form more complicated because the widget map is not exported but needed for the widget map of the customized form. We need to discuss how to solve this

fschinkel commented on eclipse-scout-core/src/table/organizer/TableOrganizerForm.ts:
```diff
@@ -0,0 +1,199 @@
+    let profileNo = 1;
+    const baseName = this.session.text('New');
+    this.profilesTable.rows.map(row => this.profilesTable.columnById('ConfigNameColumn').cellValue(row));
+    const profiles = this.profilesTable.columnById('ConfigNameColumn').cellValues();
```
remove

cguglielmo replied:
what do you mean?

fschinkel commented on eclipse-scout-core/src/table/organizer/TableOrganizerForm.ts:
```diff
@@ -0,0 +1,199 @@
+    let columnMovableToLeft = false;
+    let columnMovableToRight = false;
+    for (const column of selectedColumns) {
+      if (this.table.organizer.isColumnModifiable(column)) {
```
This does not seem correct, if one column is movable, ... then the action is enabled but only half of the expected action is executed.

cguglielmo replied:
It was intended to make it work the same way as RemoveColumn does: move only those columns that are movable. But after revisiting it I found some issues that are hard to fix. And the move helper actually wants to keep the distances between the columns when multiple columns are moved at once, this would not be guaranteed anymore. So, I simplified it.

fschinkel commented on eclipse-scout-core/src/table/organizer/TableOrganizerMenu.ts:
```diff
@@ -0,0 +1,43 @@
+    }
+    if (!this.form) {
+      this.setForm(scout.create(TableOrganizerForm, {
+        parent: this
```
move to model

cguglielmo replied:
I want to create it only when the menu is selected. If it is specified in the model it, it will be created together with the menu, see also 364072. I added a comment.

fschinkel commented on eclipse-scout-core/test/table/organizer/TableOrganizerFormSpec.ts:
```diff
@@ -0,0 +1,218 @@
+  describe('remove menu', () => {
+
+    it('is visible if a selected row can be made invisible', async () => {
+      let table = createTable(2);
```
... selected column ...

fschinkel commented on eclipse-scout-core/src/table/organizer/TableOrganizer.ts:
```diff
@@ -121,28 +114,50 @@ export class TableOrganizer implements ObjectWithType {
    */
-  hideColumn(column: Column) {
+  hideColumns(columns: Column<any>[]) {
     if (!this.table) {
```
add API change to migration guide

fschinkel commented on eclipse-scout-core/src/table/TableHeaderMenu.ts:
```diff
@@ -365,24 +371,15 @@ export class TableHeaderMenu extends Popup implements TableHeaderMenuModel {
 
     this.moveGroup.render(this.$columnActions);
-    // link buttons with the group header, the header is updated with the text of the action
-    aria.linkElementWithLabel(this.toBeginButton.$container, this.moveGroup.$text);
```
how do these aria labels work now?

cguglielmo replied:
TableHeaderMenuGroup.ts now sets the aria attributes

cguglielmo committed:
MoveTableRowMenuHelper: move only if all selected rows can be moved

If a rowFilter was set, the movable rows were moved and the others
stayed. The menus were sometimes disabled even though some of the rows
could be moved. Also, the move helper actually wants to keep the
distance between the rows when multiple rows are moved at once. This
cannot be guaranteed this way -> the rows can now only be moved if
all rows are movable.

Also
- table.moveRowsUpOrDown is based on table.rows not table.visibleRows
  -> _updateMenus must not use visibleRows either
- _updateMenus may be called by the rowOrderChanged event triggered
   when the rows are moved. This event should be ignored because other
   handlers may have to update their state first because it may
   influence the rowFilter -> just do it after moving the rows.
- renamed selectedRowsFilter to rowsFilter

cguglielmo committed:
Menu: don't open popup on mouse up again if mouse down destroyed it

Use case:
A FormMenu creates its form on select and destroys it on unselect again.
The menu calls setSelected() on mouseDown and executes regular actions
on mouseUp (click). Currently, the form would be closed on mouseDown
 and recreated on mouseUp.
To fix it, doAction() is not executed on click if mousedown already
called doAction().
